Add download location option in client settings

CarlosOlivo opened the pull request:
Also allow Jellyfin to be installed/moved to external storage

Closes #264

Maxr1998 commented on app/src/main/java/org/jellyfin/mobile/settings/SettingsFragment.kt:
```diff
@@ -90,10 +86,23 @@ class SettingsFragment : Fragment() {
             titleRes = R.string.external_player_app
             enabled = appPreferences.videoPlayerType == VideoPlayerType.EXTERNAL_PLAYER
         }
+        categoryHeader(PREF_CATEGORY_DOWNLOADS) {
+            titleRes = R.string.pref_category_downloads
+        }
+        val downloadsDirs: List<SelectionItem> = mutableListOf<SelectionItem>().apply {
+            requireContext().getDownloadsPaths().forEach {
+                add(SelectionItem(it, it, null))
+            }
+        }
+        singleChoice(Constants.PREF_DOWNLOAD_LOCATION, downloadsDirs) {
+            titleRes = R.string.pref_download_location
+            initialSelection = appPreferences.downloadLocation
+        }
```
I suppose you could move the download option to the top (and call the category something like "General")

Maxr1998 commented on app/src/main/java/org/jellyfin/mobile/utils/SystemUtils.kt:
```diff
@@ -130,3 +131,26 @@ fun Context.createMediaNotificationChannel(notificationManager: NotificationMana
+    if (isEmpty()) {
+        add(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS).absolutePath)
+    }
+    toList()
```
```suggestion
```

Maxr1998 commented on app/src/main/java/org/jellyfin/mobile/utils/SystemUtils.kt:
```diff
@@ -130,3 +131,26 @@ fun Context.createMediaNotificationChannel(notificationManager: NotificationMana
 }
+
+@Suppress("DEPRECATION")
+fun Context.getDownloadsPaths() = with(mutableListOf<String>()) {
```
```suggestion
fun Context.getDownloadsPaths(): List<String> = mutableListOf<String>().apply {
```

`toList()` unnecessarily copies the List another time, just implicitly casting it to `List` is enough (and specifying the return type is generally recommended).

Maxr1998 commented on app/src/main/res/values/strings.xml:
```diff
@@ -76,4 +76,7 @@
     <string name="external_player_system_default">System default</string>
     <string name="external_player_system_default_description">Let the system handle the player choice, some players may not be fully compatible.</string>
+
+    <string name="pref_category_downloads">Downloads</string>
```
An extra category for downloads seems a bit unnecessary, I think "General" or something similar would be preferable (as already written above).

Maxr1998 commented on app/src/main/java/org/jellyfin/mobile/AppPreferences.kt:
```diff
@@ -52,6 +54,28 @@ class AppPreferences(context: Context) {
+            val defaultStorage = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS).absolutePath
+            val savedStorage = sharedPreferences.getString(Constants.PREF_DOWNLOAD_LOCATION, null) ?: defaultStorage
+            return with(File(savedStorage)) {
+                parentFile?.run {
+                    if (exists() && isDirectory) return@with savedStorage
+                }
```
What does this check? Should it check the file itself?

CarlosOlivo replied:
Maybe? File should contain the path to an existing directory on the device.

From [getExternalFilesDirs](https://developer.android.com/reference/android/content/Context#getExternalFilesDirs(java.lang.String))

- Shared storage may not always be available, since removable media can be ejected by the user. 
- The returned path may change over time if different shared storage media is inserted, so only relative paths should be persisted.

From [getExternalStoragePublicDirectory](https://developer.android.com/reference/android/os/Environment#getExternalStoragePublicDirectory(java.lang.String))

- Returns the File path for the directory. Note that this directory may not yet exist, so you must make sure it exists.

So, check if saved storage path has a parent directory that exists, if not then return the path of the primary shared storage.

For example in **File** `/storage/70C0-1C10/Download` check if **parentFile** `/storage/70C0-1C10` exists & isDirectory, if not return defaultStorage `/storage/emulated/0/Download` (usually)

DownloadManager will create the missing subdirectory `/Download` if necessary.

Maxr1998 replied:
I see. I think checking `/storage/70C0-1C10/Download` should already work, cause if `/storage/70C0-1C10` doesn't exist, `/storage/70C0-1C10/Download` won't exist as well. Should make the code a bit cleaner.

CarlosOlivo replied:
The problem with doing it that way is that the `Download` directory [may not exist](https://developer.android.com/reference/android/os/Environment#:~:text=Note%20that%20this%20directory%20may%20not%20yet%20exist). And if the folder doesn't exist, the storage doesn't exist either.

- Without `parentFile` check (`/storage/70C0-1C10/Download`)
![sMCSQGl8o6](https://user-images.githubusercontent.com/25919231/106029433-b4d0b300-6092-11eb-8480-8347a99a9fe3.gif)

- With `parentFile` check (`/storage/70C0-1C10`)
![vlUW6ztt50](https://user-images.githubusercontent.com/25919231/106029553-d336ae80-6092-11eb-92f1-99c1342ad4f3.gif)

- Download Manager will create the missing subdirectory `/Download` if necessary.
![dYWRFUFT9g](https://user-images.githubusercontent.com/25919231/106029690-f5c8c780-6092-11eb-9066-e1ae97bd146a.gif)

Maxr1998 commented on app/src/main/java/org/jellyfin/mobile/AppPreferences.kt:
```diff
@@ -52,6 +54,28 @@ class AppPreferences(context: Context) {
+        set(value) {
+            sharedPreferences.edit {
+                with(File(value)) {
+                    parentFile?.run {
+                        if (exists() && isDirectory) putString(Constants.PREF_DOWNLOAD_LOCATION, value)
+                    }
```
Same as above.

CarlosOlivo replied:
Now that I think about it, this check is redundant, since it is the same found in the `get()`, I will remove it.

Maxr1998 reviewed (dismissed):
Should be good now, I did some refactoring and rebased onto the latest master, gonna merge once you reviewed my changes @CarlosOlivo.

Maxr1998 commented:
/azp run

CarlosOlivo commented:
I can't open the "Client Settings" fragment anymore.

```
2021-02-07 11:05:46.341 6189-6397/org.jellyfin.mobile.debug W/System.err: androidx.fragment.app.Fragment$InstantiationException: Unable to instantiate fragment org.jellyfin.mobile.settings.SettingsFragment: calling Fragment constructor caused an exception
2021-02-07 11:05:46.341 6189-6397/org.jellyfin.mobile.debug W/System.err:     at androidx.fragment.app.FragmentFactory.instantiate(FragmentFactory.java:128)
2021-02-07 11:05:46.341 6189-6397/org.jellyfin.mobile.debug W/System.err:     at org.koin.androidx.fragment.android.KoinFragmentFactory.instantiate(KoinFragmentFactory.kt:20)
2021-02-07 11:05:46.341 6189-6397/org.jellyfin.mobile.debug W/System.err:     at androidx.fragment.app.FragmentTransaction.createFragment(FragmentTransaction.java:150)
2021-02-07 11:05:46.341 6189-6397/org.jellyfin.mobile.debug W/System.err:     at androidx.fragment.app.FragmentTransaction.add(FragmentTransaction.java:213)
2021-02-07 11:05:46.341 6189-6397/org.jellyfin.mobile.debug W/System.err:     at org.jellyfin.mobile.bridge.NativeInterface.openClientSettings(NativeInterface.kt:203)
2021-02-07 11:05:46.341 6189-6397/org.jellyfin.mobile.debug W/System.err:     at android.os.MessageQueue.nativePollOnce(Native Method)
2021-02-07 11:05:46.341 6189-6397/org.jellyfin.mobile.debug W/System.err:     at android.os.MessageQueue.next(MessageQueue.java:336)
2021-02-07 11:05:46.342 6189-6397/org.jellyfin.mobile.debug W/System.err:     at android.os.Looper.loop(Looper.java:174)
2021-02-07 11:05:46.342 6189-6397/org.jellyfin.mobile.debug W/System.err:     at android.os.HandlerThread.run(HandlerThread.java:67)
2021-02-07 11:05:46.342 6189-6397/org.jellyfin.mobile.debug W/System.err: Caused by: java.lang.reflect.InvocationTargetException
2021-02-07 11:05:46.342 6189-6397/org.jellyfin.mobile.debug W/System.err:     at java.lang.reflect.Constructor.newInstance0(Native Method)
2021-02-07 11:05:46.342 6189-6397/org.jellyfin.mobile.debug W/System.err:     at java.lang.reflect.Constructor.newInstance(Constructor.java:343)
2021-02-07 11:05:46.342 6189-6397/org.jellyfin.mobile.debug W/System.err:     at androidx.fragment.app.FragmentFactory.instantiate(FragmentFactory.java:115)
2021-02-07 11:05:46.342 6189-6397/org.jellyfin.mobile.debug W/System.err: 	... 8 more
2021-02-07 11:05:46.342 6189-6397/org.jellyfin.mobile.debug W/System.err: Caused by: java.lang.IllegalStateException: Method addObserver must be called on the main thread
2021-02-07 11:05:46.342 6189-6397/org.jellyfin.mobile.debug W/System.err:     at androidx.lifecycle.LifecycleRegistry.enforceMainThreadIfNeeded(LifecycleRegistry.java:317)
2021-02-07 11:05:46.342 6189-6397/org.jellyfin.mobile.debug W/System.err:     at androidx.lifecycle.LifecycleRegistry.addObserver(LifecycleRegistry.java:172)
2021-02-07 11:05:46.342 6189-6397/org.jellyfin.mobile.debug W/System.err:     at androidx.fragment.app.Fragment.initLifecycle(Fragment.java:496)
2021-02-07 11:05:46.342 6189-6397/org.jellyfin.mobile.debug W/System.err:     at androidx.fragment.app.Fragment.<init>(Fragment.java:476)
2021-02-07 11:05:46.342 6189-6397/org.jellyfin.mobile.debug W/System.err:     at org.jellyfin.mobile.settings.SettingsFragment.<init>(SettingsFragment.kt:18)
2021-02-07 11:05:46.342 6189-6397/org.jellyfin.mobile.debug W/System.err: 	... 11 more
2021-02-07 11:05:46.344 6189-6189/org.jellyfin.mobile.debug I/chromium: [INFO:CONSOLE(66)] "Uncaught Error: Java exception was raised during method invocation", source: https://demo.jellyfin.org/native/nativeshell.js (66)
```

CarlosOlivo commented:
> Should be good now, I did some refactoring and rebased onto the latest master, gonna merge once you reviewed my changes @CarlosOlivo.

So, category "Download" > "General".
Rebased again for the fragment fix in master, LGTM!

Maxr1998 commented:
> I can't open the "Client Settings" fragment anymore.
This was introduced by the dependency updates, but yeah, fixed now.

> So, category "Download" > "General".
Yeah, I decided it's fine for now 😄 

> Rebased again for the fragment fix in master, LGTM!
Thanks, gonna merge then!
